src/common/device_helpers.cuh

```diff
@@ -866,25 +870,31 @@ class AllReducer {
    * group.
    *
    * \param device_ordinals The device ordinals.
+   * \param opg Whether to initialize AllReducer with a single process per GPU
    */
 
   void Init(const std::vector<int> &device_ordinals) {
 #ifdef XGBOOST_USE_NCCL
     /** \brief this >monitor . init. */
     this->device_ordinals = device_ordinals;
     comms.resize(device_ordinals.size());
-    dh::safe_nccl(ncclCommInitAll(comms.data(),
-                                  static_cast<int>(device_ordinals.size()),
-                                  device_ordinals.data()));
+    auto id = GetUniqueId();
+
+    for (size_t i = 0; i < device_ordinals.size(); i++) {
+      int rank = rabit::GetRank() + device_ordinals[i];
+      dh::safe_cuda(cudaSetDevice(dev));
+      dh::safe_nccl(ncclCommInitRank(
+        &(comms[i]),
+        rabit::GetWorldSize() * static_cast<int>(device_ordinals.size()),
+        id, rank));
+    }
+
     streams.resize(device_ordinals.size());
     for (size_t i = 0; i < device_ordinals.size(); i++) {
       safe_cuda(cudaSetDevice(device_ordinals[i]));
       safe_cuda(cudaStreamCreate(&streams[i]));
     }
     initialised_ = true;
-#else
-    CHECK_EQ(device_ordinals.size(), 1)
-        << "XGBoost must be compiled with NCCL to use more than one GPU.";
```

Comment:
In the case of one process per GPU, the length of `device_ordinals` will always be one.